src/Telemetry/Library/TelemetryManager.cs

```diff
@@ -0,0 +1,444 @@
+                    eventProperties.Name = eventNamePrefix + eventProperties.Name;
+                }
+
+                if (!eventProperties.Properties.ContainsKey("TimestampUTC"))
+                {
+                    eventProperties.SetProperty("TimestampUTC", DateTime.UtcNow);
+                }
+
+                telemetryLogger!.LogEvent(eventProperties);
+
+                #if DEBUG
+                LogToDebug($"{telemetryLogger!.GetType().Name} logged event {eventProperties.Name}");
+                #endif
+            });
+
+        /// <summary>
+        /// Logs an event with the given eventName, without any additional properties
+        /// Adds an event name prefix if necessary.
+        /// </summary>
+        public static void LogEvent(string eventName) =>
+            CheckAndRunSafe(() => LogEvent(new EventProperties() { Name = ValidateEventName(eventName) }));
+
+        public static void SetContext(string name, object value, TelemetryPropertyType propertyType, bool isPii) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, propertyType, isPii));
+
+        public static void SetContext(string name, string? value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void SetContext(string name, int value, bool isPii = false) =>
+            SetContext(name, (long)value, isPii);
+
+        public static void SetContext(string name, long value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void SetContext(string name, Guid value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void SetContext(string name, bool value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void SetContext(string name, double value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void SetContext(string name, DateTime value, bool isPii = false) =>
+            CheckAndRunSafe(() => telemetryLogger!.SetContext(name, value, isPii.ToPiiKind()));
+
+        public static void UploadNow() =>
+            CheckAndRunSafe(() => LogManager.UploadNow());
+
+        private static void CheckAndRunSafe(Action action, bool initializingOrTearingDown = false)
+        {
+            if (TelemetryOptOut)
+            {
+                return;
+            }
+
+            try
+            {
+                if (telemetryLogger == null && !initializingOrTearingDown)
+                {
+                    throw new InvalidOperationException("TelemetryManager has not been initialized. Please call TelemetryManager.Initialize before attempting to log anything.");

```

Comment:
not very safe if it thrown an exception, is it? :)
in this particular case it might be ok to throw one, but in general we should avoid anything that causes telemetry to create any blockers from the actual app.

Reply:
The design that I chose is for the Telemetry Library to internally throw exceptions as needed to surface unexpected errors.

But then, there is another option for exceptions to be suppressed, including this one, or other unexpected exceptions. The ReadMe provides an explanation on how it works.